pyls/rpc_manager.py

```diff
@@ -0,0 +1,171 @@
+
+    def consume_requests(self):
+        """ Infinite loop watching for messages from the client."""
+        for message in self._message_manager.get_messages():
+            if isinstance(message, JSONRPCBaseResponse):
+                self._handle_response(message)
+            else:
+                self._handle_request(message)
+
+    def _handle_request(self, request):
+        """Execute corresponding handler for the recieved request
+
+        Args:
+            request (JSONRPCBaseRequest): request to act upon
+
+        Note:
+            requests are handled asynchronously if the handler returns a callable, otherwise they are handle
+            synchronously by the main thread
+        """
+        if self._shutdown and request.method != 'exit':
+            return
+
+        params = request.params if request.params is not None else {}
+        try:
+            maybe_handler = self._message_handler(request.method, params)
+        except KeyError:
+            log.debug("No handler found for %s", request.method)
+            self._message_manager.write_message(
+                JSONRPC20Response(_id=request._id, error=JSONRPCMethodNotFound()._data).data)
+            return
+
+        if request._id in self._received_requests:
+            log.error('Received request %s with duplicate id', request.data)
+        elif callable(maybe_handler):
+            self._handle_async_request(request, maybe_handler)
+        elif not request.is_notification:
+            log.debug("sync request %s", request._id)
+            response = _make_response(request, result=maybe_handler)
+            self._message_manager.write_message(response.data)
+
+    def _handle_async_request(self, request, handler):
+        log.debug("async request %s", request._id)
+        future = self._executor_service.submit(handler)
+
+        if request.is_notification:
+            return
+
+        self._received_requests[request._id] = future
+
+        def did_finish_callback(completed_future):
+            if completed_future.cancelled():
+                log.debug('Cleared cancelled request %d', request._id)
+                del self._received_requests[request._id]
+                return
+
+            error, trace = completed_future.exception_info()
+            del self._received_requests[request._id]
+            if error is not None:
+                log.error("Failed to handle request %s with error %s %s", request._id, error, trace)
+                # TODO(forozco): add more descriptive error
+                response = _make_response(request, errror=JSONRPCInternalError()._data)
+            else:
+                response = _make_response(request, result=completed_future.result())
+            self._message_manager.write_message(response.data)
+        future.add_done_callback(did_finish_callback)
+
+    def _handle_response(self, response):
+        try:
+            request = self._sent_requests[response._id]
```

Comment:
Just guard with `if response._id not in self._sent_requests` - cleaner to read

Reply:
its more pythonic to do it this way, EAFP

Reply:
Whether you do `LBYL` or `EAFP` is only relevant for performance reasons. If this is supposed to fail 99% of the time, do `LBYL` because exception handingly 99% of the time is much more extensive than the if checking.

> In DonaldKnuth's paper "StructuredProgrammingWithGoToStatements", he wrote: "Programmers waste enormous amounts of time thinking about, or worrying about, the speed of noncritical parts of their programs, and these attempts at efficiency actually have a strong negative impact when debugging and maintenance are considered. We should forget about small efficiencies, say about 97% of the time: premature optimization is the root of all evil. Yet we should not pass up our opportunities in that critical 3%." http://wiki.c2.com/?PrematureOptimization